### PR TITLE
Auto deploy API docs 2.0

### PR DESCRIPTION
**Pull request by natke**
Fourth ~third~ ~second~ try at auto publishing API docs.

This approach generates artifacts for the different languages APIs when the sources associated with that API change, then the site build loads those artifacts when triggered by changes to the gh-pages branch.

(Hence there may a delay between a change in the files that generate the API docs and the site publish. If immediate publication is required, there is a manual trigger to build and publish the site)

- [x] Java
- [x] C#
- [x] Python
- [x] C/C++
- [x] Site build

After merge of this PR
- [ ] Remove the site build config from GitHub pages settings, as that is now done by the publish-gh-pages.yml workflow
- [ ] Remove generated API docs from source control in gh-pages 

I also removed a compliance violating Python example.  See [original PR](https://github.com/microsoft/onnxruntime/pull/15030)

Staged here: https://natke.github.io/onnxruntime/docs/api/

### CODE REVIEW

File: .github/workflows/publish-c-apidocs.yml [hunk 1]
```diff
@@ -21,27 +27,50 @@ jobs:
...
       - name: Run doxygen
         run: |
           mkdir -p build/doxygen
           cd docs/c_cxx
           ../../doxygen-1.9.6/bin/doxygen
-      - uses: actions/checkout@v2
+      - name: Log source commit
+        run: git rev-parse --short HEAD > build/doxygen/html/source-version.txt
+      - name: Download deployed site
+        uses: dawidd6/action-download-artifact@v2
```
Comment (prasanthpul) [inline .github/workflows/publish-c-apidocs.yml:38]: why this instead of https://github.com/actions/download-artifact?

Comment (natke) [reply .github/workflows/publish-c-apidocs.yml:38]: I tried that first but AFAIK it only allows artifacts generated in the same workflow.

File: .github/workflows/publish-c-apidocs.yml [hunk 2]
```diff
@@ -21,27 +27,50 @@ jobs:
...
         with:
-          ref: gh-pages
-          clean: false
-      - name: Move API docs into target area
+          name: github-pages
+          workflow: jekyll-gh-pages.yml
```
Comment (edgchen1) [inline .github/workflows/publish-c-apidocs.yml:41]: is this a new file? I didn't see it

Comment (natke) [reply .github/workflows/publish-c-apidocs.yml:41]: Good spot! That is the config change I mentioned in the PR description. I've updated the description to make this more explicit

**Review by snnn — DISMISSED** [review]

File: .github/workflows/publish-apidocs.yml [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+        run: |
+          sudo apt-get update
+          sudo apt-get install python3-pip
+          sudo apt-get install graphviz
+      - name: Python API docs - install dependencies
+        run: |
+          python3 -m pip install --upgrade pip
+          cd docs/python
+          python3 -m pip install -r requirements.txt
+          python3 -m pip install -i https://aiinfra.pkgs.visualstudio.com/PublicPackages/_packaging/ORT-Nightly/pypi/simple/ ort-nightly
```
Comment (edgchen1) [inline .github/workflows/publish-apidocs.yml:95]: why do we need to install the ORT nightly package? is the current source code not sufficient to generate the Python docs?

**Review by snnn — DISMISSED** [review]

File: .github/workflows/publish-apidocs.yml (hunk 3 above)
Comment (natke) [reply .github/workflows/publish-apidocs.yml]: That's a good question. I'll look into it (but after we get this PR merged)

File: docs/python/inference/examples_md.rst [hunk 4]
```diff
@@ -4,21 +4,20 @@
...
         :maxdepth: 1
         :caption: Contents:
-        
+
         auto_examples/plot_load_and_predict
         auto_examples/plot_common_errors
         auto_examples/plot_train_convert_predict
         auto_examples/plot_pipeline
         auto_examples/plot_backend
-        auto_examples/plot_convert_pipeline_vectorizer
```
Comment (prasanthpul) [inline docs/python/inference/examples_md.rst:22]: why is this part of this PR?

Comment (natke) [reply docs/python/inference/examples_md.rst:22]: The example was flagged as inappropriate.

Comment (natke) [reply docs/python/inference/examples_md.rst:22]: I've updated the PR description

File: .github/workflows/publish-gh-pages.yml [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+      - name: Download Python apidocs artifact
+        uses: dawidd6/action-download-artifact@v2
+        with:
+          name: onnxruntime-python-apidocs
+          workflow: publish-python-apidocs.yml
+          path: apidocs
+
+      - name: Move apidocs folder into place
+        run: |
+          sudo rm -rf _site/docs/api/c
```
Comment (edgchen1) [inline .github/workflows/publish-gh-pages.yml:73]: do we need to use `sudo`?

File: .github/workflows/publish-gh-pages.yml [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+        uses: actions/jekyll-build-pages@v1
+        with:
+          source: ./
+          destination: ./_site
+
+      - name: Download C apidocs artifact
+        uses: dawidd6/action-download-artifact@v2
+        with:
+          name: onnxruntime-c-apidocs
+          workflow: publish-c-apidocs.yml
```
Comment (edgchen1) [inline .github/workflows/publish-gh-pages.yml:47]: should we also specify "branch: main" for the download artifact steps?

File: .github/workflows/publish-gh-pages.yml [hunk 7]
```diff
@@ -0,0 +1,97 @@
+# Sample workflow for building and deploying a Jekyll site to GitHub Pages
+name: Publish site
+
+on:
+  # Runs on pushes targeting the default branch
```
Comment (edgchen1) [inline .github/workflows/publish-gh-pages.yml:5]: nit: gh-pages is not the "default branch", maybe update or remove this comment

File: .github/workflows/publish-gh-pages.yml [hunk 8]
```diff
@@ -0,0 +1,97 @@
+# Sample workflow for building and deploying a Jekyll site to GitHub Pages
+name: Publish site
+
+on:
```
Comment (edgchen1) [inline .github/workflows/publish-gh-pages.yml:4]: should we also add a scheduled trigger to periodically pick up API doc artifact updates?

File: .github/workflows/publish-c-apidocs.yml [hunk 9]
```diff
@@ -5,11 +5,17 @@ on:
       - main
     paths:
       - include/onnxruntime/core/session
-
   workflow_dispatch:
```
Comment (edgchen1) [inline .github/workflows/publish-c-apidocs.yml:8]: should we also add a scheduled trigger for the API doc workflows so the generated artifact gets refreshed and doesn't risk exceeding the retention period? though I think the retention period is 90 days by default so it doesn't need to be very frequent

File: .github/workflows/publish-c-apidocs.yml (hunk 1 above)
Comment (Soheilwikey) [reply .github/workflows/publish-c-apidocs.yml]: DNABKKOFFICJ

File: .github/workflows/publish-gh-pages.yml (hunk 5 above)
Comment (natke) [reply .github/workflows/publish-gh-pages.yml:77]: Yes, needed for both commands. I experimented with all combos here: https://github.com/natke/onnxruntime/actions/runs/4492983164/jobs/7903602815

File: .github/workflows/publish-gh-pages.yml (hunk 6 above)
Comment (natke) [reply .github/workflows/publish-gh-pages.yml:47]: Done

File: .github/workflows/publish-gh-pages.yml (hunk 7 above)
Comment (natke) [reply .github/workflows/publish-gh-pages.yml:5]: Fixed

File: .github/workflows/publish-gh-pages.yml (hunk 8 above)
Comment (natke) [reply .github/workflows/publish-gh-pages.yml:4]: I think we're good with the gh-pages branch trigger as well as the manual deploy

File: .github/workflows/publish-c-apidocs.yml (hunk 9 above)
Comment (natke) [reply .github/workflows/publish-c-apidocs.yml:8]: That's a good idea. The default is actually 1 day!

Comment (natke) [reply .github/workflows/publish-c-apidocs.yml:12]: Added

**Review by edgchen1 — APPROVED** [review]

**Review by snnn — APPROVED** [review]

**Pull request merged**